core/platform/auth/firebase_auth_services_test.py

```diff
@@ -0,0 +1,197 @@
+                auth_services.authenticate_request(request), 'auth_id')
+
+    @mock_successful_firebase_initialization
+    def test_returns_none_when_auth_header_is_missing(self):
+        request = self.make_request()
+        self.assertIsNone(auth_services.authenticate_request(request))
+
+    @mock_successful_firebase_initialization
+    def test_returns_none_when_auth_header_uses_wrong_scheme_type(self):
+        request = self.make_request(auth_header='Basic password=123')
+        self.assertIsNone(auth_services.authenticate_request(request))
+
+    @mock_successful_firebase_initialization
+    def test_returns_none_when_auth_token_is_invalid(self):
+        verify_swap = self.swap_to_always_raise(
+            firebase_admin.auth, 'verify_id_token',
+            error=firebase_exceptions.InvalidArgumentError('bad!'))
+        with verify_swap:
+            request = self.make_request(auth_header='Bearer DUMMY_JWT')
+            self.assertIsNone(auth_services.authenticate_request(request))
+
+    @mock_successful_firebase_initialization
+    def test_returns_none_when_auth_token_is_missing_subject_identifier(self):
+        mock_verify = self.swap_to_always_return(
+            firebase_admin.auth, 'verify_id_token', value={})
+        with mock_verify:
+            request = self.make_request(auth_header='Bearer DUMMY_JWT')
+            self.assertIsNone(auth_services.authenticate_request(request))
+
+
+class AuthIdUserIdAssociationOperationsTests(test_utils.GenericTestBase):
+
+    def get_associated_user_id(self, auth_id):
+        """Fetches the given associated user ID from storage manually."""
+        model = (
+            auth_models.UserIdByFirebaseAuthIdModel.get(auth_id, strict=False))
+        return None if model is None else model.user_id
+
+    def put_association(self, pair):
+        """Commits the given association to storage manually."""
+        auth_id, user_id = pair
+        auth_models.UserIdByFirebaseAuthIdModel(
+            id=auth_id, user_id=user_id).put()
+
+    def test_get_association_that_exists(self):
+        self.put_association(auth_domain.AuthIdUserIdPair('sub', 'uid'))
+
+        self.assertEqual(
+            auth_services.get_user_id_from_auth_id('sub'), 'uid')
+
+    def test_get_association_that_does_not_exist(self):
+        self.assertIsNone(
+            auth_services.get_user_id_from_auth_id('does_not_exist'))
+
+    def test_get_multi_associations_that_exist(self):
+        self.put_association(
+            auth_domain.AuthIdUserIdPair('sub1', 'uid1'))
+        self.put_association(
+            auth_domain.AuthIdUserIdPair('sub2', 'uid2'))
+        self.put_association(
+            auth_domain.AuthIdUserIdPair('sub3', 'uid3'))
+
+        self.assertEqual(
+            auth_services.get_multi_user_ids_from_auth_ids(
+                ['sub1', 'sub2', 'sub3']),
+            ['uid1', 'uid2', 'uid3'])
+
+    def test_get_multi_associations_that_do_not_exist(self):
```

Comment:
```suggestion
    def test_get_multi_associations_one_does_not_exist(self):
```

Slightly clearer since only one is nonexistent (versus all of them).

Reply:
Done.